File: bumperbot_ws/src/bumperbot_cpp_examples/src/simple_parameter.cpp
#include <rclcpp/rclcpp.hpp>
#include <rcl_interfaces/msg/set_parameters_result.hpp>
#include <string>
#include <vector>

using std::placeholders::_1;

class SimpleParameter: public rclcpp::Node {
public:
  SimpleParameter(): Node("simple_parameter") {
    declare_parameter<int>("simple_int_param", 28);
    declare_parameter<std::string>("simple_string_param", "Default value");
    add_on_set_parameters_callback(std::bind(&SimpleParameter::paramChangeCallback, this, _1));
  }

private:
  OnSetParametersCallbackHandle::SharedPtr param_callback_handle;

  rcl_interfaces::msg::SetParametersResult paramChangeCallback(const std::vector<rclcpp::Parameter> &params) {
    rcl_interfaces::msg::SetParametersResult result;
    
    // &-> ref, no copy
    for(const auto& param: params) {
      if(param.get_name() == "simple_int_param" 
      && param.get_type() == rclcpp::ParameterType::PARAMETER_INTEGER) {
        RCLCPP_INFO_STREAM(get_logger(), "Param simple_int_param changed, value = " << param.as_int());
        result.successful = true;
      }

      if(param.get_name() == "simple_string_param" 
      && param.get_type() == rclcpp::ParameterType::PARAMETER_STRING) {
        RCLCPP_INFO_STREAM(get_logger(), "Param simple_int_param changed, value = " << param.as_string());
        result.successful = true;
      }
    }

    return result;
  }
};

int main(int argc, char* argv[]) {
  rclcpp::init(argc,argv);
  auto simple_parameter = std::make_shared<SimpleParameter>();
  rclcpp::spin(simple_parameter);
  rclcpp::shutdown();

  return 0; 
}

File: bumperbot_ws/src/bumperbot_py_examples/bumperbot_py_examples/simple_parameter.py
import rclpy
from rclpy.node import Node
from rclpy.parameter import Parameter
from rcl_interfaces.msg import SetParametersResult


class SimpleParameter(Node):
  def __init__(self):
    # init with node name
    super().__init__("simple_parameter")
    # param name / default value
    self.declare_parameter("simple_int_param", 20)
    self.declare_parameter("simple_string_param", "Default Value")

    self.add_on_set_parameters_callback(self.paramChangeCallback)

  def paramChangeCallback(self, params: list[Parameter]) -> SetParametersResult:
    result = SetParametersResult()
    
    for param in params:
      if param.name == "simple_int_param" and param.type_ == Parameter.Type.INTEGER:
        self.get_logger().info(f"param simple_int_param changed to {param.value}")
        result.successful = True

      elif param.name == "simple_string_param" and param.type_ == Parameter.Type.STRING:
        self.get_logger().info(f"param simple_int_param changed to {param.value}")
        result.successful = True

    return result
  
def main():
  rclpy.init()
  simple_parameter = SimpleParameter()
  rclpy.spin(simple_parameter)
  simple_parameter.destroy_node()
  rclpy.shutdown()
  
if __name__ == "__main__":
  main()